cryoet_data_portal_neuroglancer/precompute/contrast_limits.py

```diff
@@ -0,0 +1,557 @@
+                "high_percentile": {"type": "randint", "args": [50, 100]},
+            },
+        )
+
+    def contrast_limits_from_mean(
+        self,
+        multipler: float = 3.0,
+    ) -> tuple[float, float]:
+        """Calculate the contrast limits from the mean and RMS.
+
+        Parameters
+        ----------
+            multipler: float, optional.
+                The multiplier for the RMS value.
+                By default 3.0.
+
+        Returns
+        -------
+            tuple[float, float]
+                The calculated contrast limits.
+        """
+        mean_value = np.mean(self.volume)
+        rms_value = np.sqrt(np.mean(self.volume**2))
+        width = multipler * rms_value
+
+        return mean_value - width, mean_value + width
+
+
+class GMMContrastLimitCalculator(ContrastLimitCalculator):
+
+    def compute_contrast_limit(
+        self,
+        low_variance_mult: float = 3.0,
+        high_variance_mult: float = 0.5,
+        max_components: int = 3,
+    ) -> tuple[float, float]:
+        """Calculate the contrast limits using Gaussian Mixture Model.
+
+        Parameters
+        ----------
+        max_components: int, optional.
+            The max number of components to use for the GMM.
+            By default 5.
+        low_variance_mult: float, optional.
+            The multiplier for the low variance.
+            By default 3.0.
+        high_variance_mult: float, optional.
+            The multiplier for the high variance.
+            By default 0.5.
+
+        Returns
+        -------
+            tuple[float, float]
+                The calculated contrast limits.
+        """
+        covariance_type = "full"
+        sample_data = self.volume.flatten()
+
+        # Find the best number of components - using BIC
+        # BIC is a criterion for model selection among a finite set of models
+        # The model with the lowest BIC is preferred.
+        bics = np.zeros(shape=(max_components, 2))
+        for n in range(1, max_components + 1):
+            n = int(n)
+            gmm = GaussianMixture(
+                n_components=n,
+                covariance_type=covariance_type,
+                max_iter=100,
+                random_state=42,
+                init_params="k-means++",
+            )
+            try:
+                gmm.fit(sample_data.reshape(-1, 1))
+            except ValueError:
+                bics[n - 1] = [np.inf, n]
+            else:
+                bics[n - 1] = [gmm.bic(sample_data.reshape(-1, 1)), n]
+        min_bic_index = np.argmin(bics[:, 0])
+        best_n = int(bics[min_bic_index, 1])
+
+        # With less components, we need to be more conservative
+        # Hence the standard_deviation multiplier is higher
+        std_dev_multi_dict = {
+            1: (2.0, 0.5),
+            2: (2.2, 0.65),
+            3: (3.0, 0.8),
+        }
+
+        # Redo the best with more iterations
+        self.gmm_estimator = GaussianMixture(
```

Comment:
Depending on the road we take (using internal states, or not), `self.gmm_estimator` should be considered as a simple variable, and not as an instance variable.